pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,414 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/golang/glog"
+)
+
+const (
+	nodeUpdateChannelSize = 10000
+	podUpdateChannelSize  = 50000
+	retries               = 4
+)
+
+func computeTaintDifference(left []v1.Taint, right []v1.Taint) []v1.Taint {
+	result := []v1.Taint{}
+	for i := range left {
+		found := false
+		for j := range right {
+			if left[i] == right[j] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			result = append(result, left[i])
+		}
+	}
+	return result
+}
+
+// copy of 'computeTaintDifference' - long live lack of generics...
+func computeTolerationDifference(left []v1.Toleration, right []v1.Toleration) []v1.Toleration {
+	result := []v1.Toleration{}
+	for i := range left {
+		found := false
+		for j := range right {
+			if left[i] == right[j] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			result = append(result, left[i])
+		}
+	}
+	return result
+}
+
+type nodeUpdateItem struct {
+	oldNode   *v1.Node
+	newNode   *v1.Node
+	oldTaints []v1.Taint
+	newTaints []v1.Taint
+}
+
+type podUpdateItem struct {
+	oldPod         *v1.Pod
+	newPod         *v1.Pod
+	oldTolerations []v1.Toleration
+	newTolerations []v1.Toleration
+}
+
+// NoExecuteTaint manager listens to Taint/Toleration changes and is resposible for removing Pods
+// from Nodes tainted with NoExecute Taints.
+type NoExecuteTaintManager struct {
+	client             clientset.Interface
+	taintEvictionQueue *TimedWorkerQueue
+	// keeps a map from nodeName to all noExecute taints on that Node
+	taintedNodes     map[string][]v1.Taint
+	taintedNodesLock sync.Mutex
+
+	nodeUpdateChannel chan nodeUpdateItem
+	podUpdateChannel  chan podUpdateItem
+}
+
+func deletePodHandler(c clientset.Interface) func(args *WorkArgs) {
+	return func(args *WorkArgs) {
+		ns := args.NamespacedName.Namespace
+		name := args.NamespacedName.Name
+		glog.V(0).Infof("NoExecuteTaintManager is deleting Pod: %v", args.NamespacedName.String())
+		c.Core().Pods(ns).Delete(name, &metav1.DeleteOptions{})
+	}
+}
+
+func getNonExecuteTaints(taints []v1.Taint) []v1.Taint {
+	result := []v1.Taint{}
+	for i := range taints {
+		if taints[i].Effect == v1.TaintEffectNoExecute {
+			result = append(result, taints[i])
+		}
+	}
+	return result
+}
+
+// Returns true and list of Tolerations matching all Taints if all are tolerated, or false otherwise.
+func getMatchingTolerations(taints []v1.Taint, tolerations []v1.Toleration) (bool, []v1.Toleration) {
```

Comment:
This seems to be a generic enough helper. I think that function like this should leave somewhere else (or at least in some util file) - this is not specific to the controller at all.